scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -1919,30 +2179,117 @@ void BPF_STRUCT_OPS(lavd_enqueue, struct task_struct *p, u64 enq_flags)
 	put_global_rq(p, taskc, cpuc, enq_flags);
 }
 
+static bool is_kernel_task(struct task_struct *p)
+{
+	return p->mm == NULL;
+}
+
 void BPF_STRUCT_OPS(lavd_dispatch, s32 cpu, struct task_struct *prev)
 {
+	struct bpf_cpumask *active, *ovrflw;
+	struct task_struct *p;
+
+	bpf_rcu_read_lock();
+
+	/*
+	 * Prepare cpumasks.
+	 */
+	active = active_cpumask;
+	ovrflw = ovrflw_cpumask;
+	if (!active || !ovrflw) {
+		scx_bpf_error("Failed to prepare cpumasks.");
+		goto unlock_out;
+	}
+
+	/*
+	 * If the CPU belonges to the active or overflow set, dispatch a task.
+	 */
+	if (bpf_cpumask_test_cpu(cpu, cast_mask(active)) ||
+	    bpf_cpumask_test_cpu(cpu, cast_mask(ovrflw))) {
+		scx_bpf_consume(LAVD_GLOBAL_DSQ);
+		goto unlock_out;
+	}
+
 	/*
-	 * Now, a CPU has no task to run, so it gets a task from the global run
-	 * queue for execution.
+	 * If this CPU is not either in active or overflow CPUs, it tries to
+	 * find and run a task pinned to run on this CPU.
 	 */
-	scx_bpf_consume(LAVD_GLOBAL_DSQ);
+	__COMPAT_DSQ_FOR_EACH(p, LAVD_GLOBAL_DSQ, 0) {
+		/*
+		 * Prioritize kernel tasks because most kernel tasks are pinned
+		 * to a paritular CPU and latency-critical (e.g., ksoftirqd,
+		 * kworker, etc).
+		 */
+		if (is_kernel_task(p)) {
+			scx_bpf_consume(LAVD_GLOBAL_DSQ);
+			break;
+		}
+
+		/*
+		 * This is a hack to bypass the restriction of the current BPF
+		 * not trusting the pointer p. Once the BPF verifier gets
+		 * smarter, we can remove bpf_task_from_pid().
+		 */
+		p = bpf_task_from_pid(p->pid);
+		if (!p)
+			goto unlock_out;
+
+		/*
+		 * If a task can run on active or overflow CPUs, it just does
+		 * nothing to go idle. 
+		 */
+		if (bpf_cpumask_intersects(cast_mask(active), p->cpus_ptr))
+			goto release_break;
+		if (bpf_cpumask_intersects(cast_mask(ovrflw), p->cpus_ptr))
+			goto release_break;
+
+		/*
+		 * Otherwise, that means there is a task that should run on
+		 * this particular CPU. So, consume one of such tasks.
+		 *
+		 * Note that this path is not optimized since scx_bpf_consume()
+		 * should traverse until it finds any task that can run on this
+		 * CPU. The scheduled task might be runnable on the active
+		 * cores. We will optimize this path after introducing per-core
+		 * DSQ.
+		 */
+		scx_bpf_consume(LAVD_GLOBAL_DSQ);
+
+		/*
+		 * This is the first time a particular pinned user-space task
+		 * is run on this CPU at this interval. From now on, this CPU
+		 * will be part of the activee CPU so can be used to run the
+		 * pinned task and the other tasks.
+		 */
+		bpf_cpumask_test_and_set_cpu(cpu, active);
+		scx_bpf_kick_cpu(cpu, __COMPAT_SCX_KICK_IDLE);
```

Comment:
Should we check the return value of `scx_bpf_consume()` to see if we were able to successfully consume a task before promoting this core to active?

Reply:
I wasn't sure if there is a case `cpu` is not the current CPU. If so, we don't need a kick. Will fix it.

Reply:
Fixed.

Reply:
>I wasn't sure if there is a case cpu is not the current CPU. If so, we don't need a kick. Will fix it.

Ah, sorry for the confusion but I just rechecked and apparently we can invoke ops.dispatch() for remote cores for core sched.